igvc_platform/src/imu/main.cpp

```diff
@@ -71,32 +62,58 @@ int main(int argc, char** argv)
     ROS_ERROR_STREAM("failed to send message");
   }
 
-  freespace_message response;
-  ret = freespace_readMessage(ids[0], &response, 1000);
-  if (ret != FREESPACE_SUCCESS)
-  {
-    if (ret == FREESPACE_ERROR_TIMEOUT)
-    {
-      ROS_ERROR_STREAM("failed to read IMU TIMEOUT");
-    }
-    else
-    {
-      ROS_ERROR_STREAM("failed to read IMU" << ret);
-    }
-  }
-  if (response.messageType == FREESPACE_MESSAGE_DATAMODECONTROLV2RESPONSE)
-  {
-    ROS_INFO_STREAM("got response" << response.dataModeControlV2Response.modeActual);
+  return ret;
+}
+
+int main(int argc, char** argv)
+{
+  ros::init(argc, argv, "imu");
+
+  ros::NodeHandle nh;
+  ros::NodeHandle pNh("~");
+
+  // Get matrix to correct IMU orientation.
+  std::vector<double> imu_orientation_correction;
+  nh.getParam("imu_orientation_correction", imu_orientation_correction);
+
+  // Get threshold value for triggering a warning when the deviation between the filtered
+  // heading and magnetometer heading becomes too large.
+  double heading_dev_thresh_deg;
+  nh.getParam("heading_dev_thresh_deg", heading_dev_thresh_deg);
+
+  // There should be 9 elements (3x3 matrix).
+  if (imu_orientation_correction.size() != 9) {
+    ROS_ERROR_STREAM("IMU orientation correction matrix does not have 9 elements (3x3).");
+    return 0;
   }
-  else
-  {
-    ROS_ERROR_STREAM("got invalid type " << response.messageType);
+
+  // Populate an Eigen matrix.
+  Eigen::Matrix3d correction_mat(imu_orientation_correction.data());
+
+  // Publish imu data without gravity for EKF localization.
+  ros::Publisher imu_pub = nh.advertise<sensor_msgs::Imu>("/imu", 1000);
+
+  // Publish imu data with gravity for factor graph implementation.
+  ros::Publisher imu_raw_pub = nh.advertise<sensor_msgs::Imu>("/imu_raw", 1000);
+
+  // sequence of published messages - should be monotonically increasing.
+  int seq = 0;
+
+  // Initialize IMU.
+  FreespaceDeviceId ids[1];
+  int ret = initializeIMU(ids);
+  if (ret != FREESPACE_SUCCESS) {
+    return 0;
```

Comment:
I would print an error here instead of quietly ending.

Reply:
Done